### ISSUE TITLE
Handshake with failed after session was established! Evicted!

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#1284]
With Cf-2.2.2 we faced some : 

```
[ection-Handler-9] o.e.c.scandium.DTLSConnector     Handshake with [/223.224.72.128:35546] failed after session was established!
java.lang.RuntimeException: Evicted!
 at org.eclipse.californium.scandium.dtls.InMemoryConnectionStore$1$1.run(InMemoryConnectionStore.java:177) [av-lwm2m-bs.jar:na]
 at org.eclipse.californium.elements.util.SerialExecutor$1.run(SerialExecutor.java:276) [av-lwm2m-bs.jar:na]
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_212]
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_212]
 at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212]
```
or
```
[ection-Handler-8] o.e.c.scandium.DTLSConnector     Handshake with [null] failed after session was established!
java.lang.RuntimeException: Evicted!
 at org.eclipse.californium.scandium.dtls.InMemoryConnectionStore$1$1.run(InMemoryConnectionStore.java:177) [av-lwm2m-bs.jar:na]
 at org.eclipse.californium.elements.util.SerialExecutor$1.run(SerialExecutor.java:276) [av-lwm2m-bs.jar:na]
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_212]
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_212]
 at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212
```

I understand this happens when a Handshake is finished but we never get any APPLICATION_DATA (so handshake is not complete) and finally the connection is evicted.

I think this should not be a `warn` ? :thinking: Just a debug to search "bad device behavior"

We faced this a lot ~600/700 per hours.

I do a quick capture and look at the first 450 conversation and I find several handshakes (~10) which finished like this :
```
Client Hello
Hello Verify Request
Client Hello
Server Hello, Server Hello Done
Client Key Exchange, Change Cipher Spec, Encrypted Handshake Message
Change Cipher Spec, Encrypted Handshake Message
```
This looks like a device issue as we never get APPLICATION_DATA.

But I also see (2 only) :
```
Client Hello
Hello Verify Request
Client Hello
Server Hello, Server Hello Done
Client Key Exchange, Change Cipher Spec, Encrypted Handshake Message
Change Cipher Spec, Encrypted Handshake Message
Client Key Exchange, Change Cipher Spec, Encrypted Handshake Message
Client Key Exchange, Change Cipher Spec, Encrypted Handshake Message
Client Key Exchange, Change Cipher Spec, Encrypted Handshake Message
```
Like if server does not answer to the device retransmission. I tried to reproduce locally but didn't succeed.

(It would be great to have connection identity when possible in logs)

**Comment by boaks** [comment]
> But I also see (2 only) :
Maybe the client is a little more buggy and resend the records with old record sequence numbers.

**Comment by boaks** [comment]
And yes, if you want, change the log level to debug.

**Comment by boaks** [comment]
> (It would be great to have connection identity when possible in logs)

`Connection.toString()`? Something else? Here the address:port is included, so do you have something else in mind?

**Comment by boaks** [comment]
By the way, I had sometimes ago the idea, to replace a lot of the `InetSocketAddress` passed around for the message decoding/encoding with the `Connection`. Some places may be left, because there is no "connection" (early handshake, decode record). But that would cause many changes so I postponed the idea to 3.0 .

**Comment by sbernard31** [comment]
> Maybe the client is a little more buggy and resend the records with old record sequence numbers.

Good catch  🎯 !
Buggy in a funny way :joy: !
Sequence numbers of Client Key Exchange, Change Cipher Spec, Encrypted Handshake Message
- 4,5,6
- 7,5,6
- 8,5,6
- 9,5,6

**Comment by sbernard31** [comment]
> And yes, if you want, change the log level to debug.

I will do that on monday and use  `handshaker.getPeerAddress()`

> Connection.toString()? Something else? Here the address:port is included, so do you have something else in mind?

Yep ideally when we see something wrong with a peer we want to identify it. To know which kind of device is buggy and eventually report the bug. The pertinent information for us is the Sender Identity (e.g. PSK ID)

**Comment by boaks** [comment]
> The pertinent information for us is the foreign peer Identity (e.g. PSK ID)

That seems to request a customizable "Connection to logging string" function.

**Comment by sbernard31** [comment]
You think that adding the identifier of the peer in the connection.toString is not enough ?

**Comment by boaks** [comment]
I guess, that "everybody" have it's own preferences.

**Comment by sbernard31** [comment]
Sure, but I guess with peer address + conId + sessionId + peerIdentity we cover most of them.

**Comment by sbernard31** [comment]
It should be fixed by  #1286 and #1290.

For the "print identity" in logs stuff, I would create a dedicated issue.
